src/server/pps/server/poller.go

```diff
@@ -176,25 +172,27 @@ func (m *ppsMaster) pollPipelines(ctx context.Context) {
 // to CRASHING
 func (m *ppsMaster) pollPipelinePods(ctx context.Context) {
 	if err := backoff.RetryUntilCancel(ctx, backoff.MustLoop(func() error {
-		kubePipelineWatch, err := m.a.env.KubeClient.CoreV1().Pods(m.a.namespace).Watch(
-			ctx,
-			metav1.ListOptions{
-				LabelSelector: metav1.FormatLabelSelector(metav1.SetAsLabelSelector(
-					map[string]string{
-						"component": "worker",
-					})),
-				Watch: true,
-			})
+		watch, cancel, err := m.kd.WatchPipelinePods(ctx)
 		if err != nil {
 			return errors.Wrap(err, "failed to watch kubernetes pods")
 		}
-		defer kubePipelineWatch.Stop()
+		defer cancel()
```

Comment:
weird technical go question here I'm going to have to look up: when is the function in a defer bound, exactly? It seems like the closure must be created at the `defer` line, so we will keep putting `defer`s on the stack which refer to old `watch`es. Probably not an issue, but if this wasn't how go worked and instead overwriting `cancel` changed it, this would be an issue as `cancel` could end up nil.